File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(sock: socket.socket, size):
    total_read = b""

    while len(total_read) < size:
        read = sock.recv(size - len(total_read))

        if read == b"":
            raise RuntimeError("Connection closed before receiving all data")

        total_read += read

    return total_read


def send_all(sock: socket.socket, data: bytes):
    total_sent = 0

    while total_sent < len(data):
        sent = sock.send(data[total_sent:])
        total_sent += sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	totalSent := 0

	for totalSent < len(bytes) {
		sent, err := socket.Write(bytes[totalSent:])
		totalSent += sent

		if err != nil {
			return err
		}
	}

	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalRead := 0

	for totalRead < size {
		read, err := socket.Read(buff[totalRead:])
		totalRead += read

		if totalRead == size {
			return buff, nil
		}

		if err != nil {
			return nil, err
		}
	}

	return buff, nil
}
